crates/primitives-currency-swap/src/lib.rs

```diff
@@ -3,15 +3,20 @@
 // Either generate code at stadard mode, or `no_std`, based on the `std` feature presence.
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use frame_support::{sp_runtime::DispatchError, traits::Currency};
+use frame_support::{
+    sp_runtime::DispatchError,
+    traits::{fungible::Inspect, Currency},
+};
 
 /// Currency swap interface.
 pub trait CurrencySwap<AccountIdFrom, AccountIdTo> {
     /// The currency to convert from.
-    type From: Currency<AccountIdFrom>;
+    type From: Currency<AccountIdFrom>
+        + Inspect<AccountIdFrom, Balance = <Self::From as Currency<AccountIdFrom>>::Balance>;
```

Comment:
Why?

Reply:
To use `can_deposit`. `Balances` and `Evmbalances` implements it.

Reply:
I see, the question is why do we need a manual extra check if we are about to actually do the deposit in the same flow